### ISSUE TITLE
Document the App Creation Requirements?

### ISSUE THREAD

**Issue by grumple-unithera** [ggpwnkthx/Microsoft-Graph-SMTP-Relay#1]
I have been looking for this, and glad I found this. However, I will try and figure it out but was curious if you could toss some documentation together on setting up the APP that's required.

**Comment by ggpwnkthx** [comment]
Generally speaking this should work just like a regular SMTP service. However, it accepts connections with no authentication as well as any LOGIN or PLAIN authentication without TLS or SSL. It is purposefully designed to be insecure in this way so that your other services can route the messages with little delay. 

So it's **VERY** important to make sure you air-gap the inbound networking for this service and never expose its SMTP port outside of your internal container network.

Are you using a Docker or Kubernetes environment? If you can provide the YAML file you're using (with sensitive data scrubbed) I can take a look at how this would fit in.

**Comment by ggpwnkthx** [comment]
Sorry, coffee hasn't kicked in yet. I just realized you meant how the registered app in AAD/Entra should be set up. Is that correct?

**Comment by grumple-unithera** [comment]
Yeah, I think I figured that part out, but having a different issue. Trying to figure it out now.

```
microsoft-graph-smtp-relay  | Traceback (most recent call last):
microsoft-graph-smtp-relay  |   File "/usr/local/lib/python3.12/site-packages/aiosmtpd/smtp.py", line 741, in _handle_client
microsoft-graph-smtp-relay  |     await method(arg)
microsoft-graph-smtp-relay  |   File "/usr/local/lib/python3.12/site-packages/aiosmtpd/smtp.py", line 1460, in smtp_DATA
microsoft-graph-smtp-relay  |     status = await self._call_handler_hook('DATA')
microsoft-graph-smtp-relay  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
microsoft-graph-smtp-relay  |   File "/usr/local/lib/python3.12/site-packages/aiosmtpd/smtp.py", line 473, in _call_handler_hook
microsoft-graph-smtp-relay  |     status = await hook(self, self.session, self.envelope, *args)
microsoft-graph-smtp-relay  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
microsoft-graph-smtp-relay  |   File "/usr/src/app/handlers/microsoft_graph.py", line 119, in handle_DATA
microsoft-graph-smtp-relay  |     + self.app.acquire_token_for_client(scopes=[".default"])["access_token"]
microsoft-graph-smtp-relay  |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
microsoft-graph-smtp-relay  | KeyError: 'access_token'

```
I'm running this in docker (using a docker compose, I modified it a bit)

```
version: "3"
services:
  smtp_relay:
    image: microsoft-graph-smtp-relay
    container_name: microsoft-graph-smtp-relay
    environment:
      AUTHORITY: "https://login.microsoftonline.com/xxxxxxxxxxxxx" # UPDATE THIS
      CLIENT_ID: "yyyyyyyyyyyyyyyyyyyyyy" # UPDATE THIS
      CLIENT_SECRET: "zzzzzzzzzzzzzzz" # UPDATE THIS
      SMTP_RELAY_HOSTNAME: "0.0.0.0"
      SMTP_RELAY_PORT: "25"
    build:
      context: ./
      dockerfile: dockerfile
    ports:
      - "25:25"

```

**Comment by grumple-unithera** [comment]
This is going to be used on an EC2 instance to relay mail from printers in multiple offices (they are all connected via IPSec tunnels to AWS). It will all be IP locked (I only have it allowing 1 printer at the moment that I am testing from).

**Comment by grumple-unithera** [comment]
FYI this is the JSON that's being logged (I turned on debug logging, had to alter some code to get that working as it wants to write to a path it can't in the container). I changed the address from what I'm testing to test@test.com.

` {'message': {'subject': 'Test Email', 'body': {'contentType': 'text', 'content': ''}, 'toRecipients': [{'emailAddress': {'address': '<test@test.com>'}}], 'ccRecipients': [], 'bccRecipients': [], 'attachments': [{'@odata.type': '#microsoft.graph.fileAttachment', 'name': None, 'contentType': 'text/plain', 'contentBytes': 'VGVzdCBtYWlsIHRvIHZhbGlkYXRlIHNtdHBzZXJ2ZXI=', 'isInline': True, 'contentId': ''}]}, 'saveToSentItems': 'false'}
`

**Comment by ggpwnkthx** [comment]
> This is going to be used on an EC2 instance to relay mail from printers in multiple offices (they are all connected via IPSec tunnels to AWS). It will all be IP locked (I only have it allowing 1 printer at the moment that I am testing from).

Sounds like you've got your bases covered. I just don't want to understate the need for isolating the relay since it would be extremely easy to abuse it if it's not protected.

**Comment by grumple-unithera** [comment]
Oh I'm aware, the app I generated allows it to send as anyone as well. Definitely aware of the security implications of this.

**Comment by grumple-unithera** [comment]
The email it's sending is super simple. I'm attaching it (edited the FROM and the TO addresses, there's really nothing else to the email)

[7e492c9ee94c46b48b0f2a0340b5add7.zip](https://github.com/ggpwnkthx/Microsoft-Graph-SMTP-Relay/files/14195457/7e492c9ee94c46b48b0f2a0340b5add7.zip)

**Comment by grumple-unithera** [comment]
The issue (ignoring that it's probably parsing it wrong all together) that it's throwing up is the 'name': None in the JSON. I added code to just toss in a random name ('blah') if it was empty and that fixed it.

**Comment by grumple-unithera** [comment]
I also had to add some code to run regular expressions on the address lists (as email will typically come in in the form of "First Last" <flast@domain.com>" and it was just straight up tossing that into the JSON which MS threw up on. So I added code to iterate those lists and look for actual email addresses and just add those to new lists (tos, ccs, and bccs). Fixed my issue I was having with the printer sending from the address book.
